[Refactor] Convert Notebook to Scripts

agpeshal opened the pull request:
Clean-up old code with a better overall structure

agpeshal committed:
ignore cache and data files

agpeshal committed:
add adversarial class

agpeshal committed:
add dataloader

agpeshal committed:
define model architecture

agpeshal committed:
clone org image before modifying

agpeshal committed:
add attack script on CIFAR-10

agpeshal committed:
remove redundant dataloader script

agpeshal committed:
refactor attacker class'

agpeshal committed:
ignore dataset folder

agpeshal committed:
add adversarial trainer

agpeshal committed:
add train script

agpeshal committed:
re-order imports

agpeshal committed:
move public methods before private

agpeshal committed:
reformat code with black

agpeshal committed:
add training progress bar

agpeshal committed:
add pre-commit config

agpeshal committed:
add requirements

agpeshal committed:
remove old notebooks

agpeshal committed:
allow model ckpt loading

agpeshal committed:
update README

agpeshal committed:
add type hints

agpeshal committed:
remove redudant device arg from attacks

agpeshal committed:
add docstring

agpeshal committed:
rename source folder to src

abhiagg92 commented on README.md:
```diff
@@ -2,66 +2,41 @@
-cd Attack
-python cifar_model.py
-jupyter notebook
+cd source
```
```suggestion
cd src
```

agpeshal replied:
fixed.

abhiagg92 commented on README.md:
````diff
@@ -2,66 +2,41 @@
-
 ```bash
-python eval.py --defense --num_epochs 20
+cd source
````
```suggestion
cd src
```

abhiagg92 commented on requirements.txt:
```diff
@@ -0,0 +1,98 @@
+traitlets=5.14.2=pyhd8ed1ab_0
+types-tqdm=4.66.0.20240106=pypi_0
+typing-extensions=4.9.0=py310h06a4308_1
+typing_extensions=4.9.0=py310h06a4308_1
```
These two are exactly same, maybe we can remove one from the file

abhiagg92 commented on src/attack.py:
```diff
@@ -0,0 +1,69 @@
+from torch.utils.data import DataLoader
+from torchvision import datasets, transforms
+
+from .attackers import PGD
```
I generally prefer absolute imports especially if import path is short. This will make it more readable
https://stackoverflow.com/questions/4209641/absolute-vs-explicit-relative-import-of-python-module

abhiagg92 commented on src/attack.py:
```diff
@@ -0,0 +1,69 @@
+    )
+
+
+def main() -> None:
```
Maybe we can `attack` as function name for better readability.

agpeshal replied:
I feel the name of the script suffices.

abhiagg92 commented on src/attack.py:
```diff
@@ -0,0 +1,69 @@
+
+    model = CNNClassifier().to(device)
+
+    if args.ckpt is not None:
```
Will the attack work if checkpoint path is not given?

abhiagg92 commented on src/attack.py:
```diff
@@ -0,0 +1,69 @@
+
+def parse_args():
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--ckpt", type=str, default=None)
```
Maybe add description of these arguments

abhiagg92 commented on src/adversarial_trainer.py:
```diff
@@ -0,0 +1,80 @@
+            correct += (labels == predictions).sum().cpu().numpy()
+            total += len(labels)
+
+        print("Test accuracy: {:.2f}".format(correct / total * 100.0))
```
Use f-formatted print statement wherever possible https://docs.python.org/3/whatsnew/3.6.html#whatsnew36-pep498

abhiagg92 commented on src/adversarial_trainer.py:
```diff
@@ -0,0 +1,80 @@
+
+        print("Test accuracy: {:.2f}".format(correct / total * 100.0))
+
+    def predict_(self, images: torch.Tensor) -> torch.Tensor:
```
Use `_predict` instead of `predict_` as per the standard conventions to represent private method in python

abhiagg92 commented on src/adversarial_trainer.py:
```diff
@@ -0,0 +1,80 @@
+    def predict_(self, images: torch.Tensor) -> torch.Tensor:
+        return torch.argmax(self.model(images).data, dim=1)
+
+    def train_epoch_(self) -> None:
```
same here

abhiagg92 commented on src/train.py:
```diff
@@ -0,0 +1,79 @@
+    return parser.parse_args()
+
+
+def main() -> None:
```
You can also rename this method to be more descriptive

agpeshal replied:
same

abhiagg92 commented on src/train.py:
```diff
@@ -0,0 +1,79 @@
+    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+
+    batch_size = args.bs
+    train_loader = DataLoader(
```
create method to create both train and test dataloaders without code duplication

agpeshal committed:
update path in README

agpeshal committed:
remove duplicate requirements

agpeshal committed:
Restructure folder

1. Create module core with attackers and adversarial trainer
2. rename net to models with init

agpeshal committed:
fix methods of private methods

agpeshal committed:
switch f-format

agpeshal committed:
move dataloader to a script
